cli/tools/registry/pm.rs

```diff
@@ -38,313 +28,210 @@ use crate::file_fetcher::FileFetcher;
-  imports: IndexMap<String, String>,
+#[derive(Debug, Copy, Clone)]
+enum ConfigKind {
+  DenoJson,
+  PackageJson,
 }
 
-fn deno_json_imports(
-  config: &deno_config::deno_json::ConfigFile,
-) -> Result<IndexMap<String, String>, AnyError> {
-  Ok(
-    config
-      .json
-      .imports
-      .clone()
-      .map(|imports| {
-        serde_json::from_value(imports)
-          .map_err(|err| anyhow!("Malformed \"imports\" configuration: {err}"))
-      })
-      .transpose()?
-      .unwrap_or_default(),
-  )
+struct ConfigUpdater {
+  kind: ConfigKind,
+  cst: CstRootNode,
+  root_object: CstObject,
+  path: PathBuf,
+  modified: bool,
 }
-impl DenoConfig {
-  fn from_options(options: &CliOptions) -> Result<Option<Self>, AnyError> {
-    let start_dir = &options.start_dir;
-    if let Some(config) = start_dir.maybe_deno_json() {
-      Ok(Some(Self {
-        imports: deno_json_imports(config)?,
-        config: config.clone(),
-        format: DenoConfigFormat::from_specifier(&config.specifier)?,
-      }))
-    } else {
-      Ok(None)
-    }
-  }
 
-  fn add(&mut self, selected: SelectedPackage) {
-    self.imports.insert(
-      selected.import_name,
-      format!("{}@{}", selected.package_name, selected.version_req),
-    );
+impl ConfigUpdater {
+  fn new(
+    kind: ConfigKind,
+    config_file_path: PathBuf,
+  ) -> Result<Self, AnyError> {
+    let config_file_contents = std::fs::read_to_string(&config_file_path)
+      .with_context(|| {
+        format!("Reading config file '{}'", config_file_path.display())
+      })?;
+    let cst = CstRootNode::parse(&config_file_contents, &Default::default())
+      .with_context(|| {
+        format!("Parsing config file '{}'", config_file_path.display())
+      })?;
+    let root_object = cst.object_value_or_set();
+    Ok(Self {
+      kind,
+      cst,
+      root_object,
+      path: config_file_path,
+      modified: false,
+    })
   }
 
-  fn remove(&mut self, package: &str) -> bool {
-    self.imports.shift_remove(package).is_some()
+  fn display_path(&self) -> String {
+    deno_path_util::url_from_file_path(&self.path)
+      .map(|u| u.to_string())
+      .unwrap_or_else(|_| self.path.display().to_string())
   }
 
-  fn take_import_fields(
-    &mut self,
-  ) -> Vec<(&'static str, IndexMap<String, String>)> {
-    vec![("imports", std::mem::take(&mut self.imports))]
+  fn obj(&self) -> &CstObject {
+    &self.root_object
   }
-}
 
-impl NpmConfig {
-  fn from_options(options: &CliOptions) -> Result<Option<Self>, AnyError> {
-    let start_dir = &options.start_dir;
-    if let Some(pkg_json) = start_dir.maybe_pkg_json() {
-      Ok(Some(Self {
-        dependencies: pkg_json.dependencies.clone().unwrap_or_default(),
-        dev_dependencies: pkg_json.dev_dependencies.clone().unwrap_or_default(),
-        config: pkg_json.clone(),
-        fmt_options: None,
-      }))
-    } else {
-      Ok(None)
-    }
+  fn contents(&self) -> String {
+    self.cst.to_string()
```

Comment:
Does it need to be accessed more than one time? If not, consider making it consume `self` to better communicate that?

Reply:
`self.path` is still used after this.